service/worker/batcher/activities.go

```diff
@@ -41,6 +41,223 @@ var (
+// nolint:revive,cognitive-complexity
+func (a *activities) processWorkflowsWithProactiveFetching(
+	ctx context.Context,
+	config batchProcessorConfig,
+	startWorkerProcessor batchWorkerProcessor,
+	sdkClient sdkclient.Client,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+	hbd HeartBeatDetails,
+) (HeartBeatDetails, error) {
+	rateLimit := rate.Limit(config.rps)
+	burstLimit := int(math.Ceil(config.rps)) // should never be zero because everything would be rejected
+	rateLimiter := rate.NewLimiter(rateLimit, burstLimit)
+
+	concurrency := int(math.Max(1, float64(config.concurrency)))
+
+	taskCh := make(chan task, concurrency)
+	respCh := make(chan taskResponse, concurrency)
+
+	// Ticker for frequent heartbeats to avoid timeout during slow processing, 1/4 of the default heartbeat timeout (10s)
+	heartbeatTicker := time.NewTicker(defaultActivityHeartBeatTimeout / 4)
+	defer heartbeatTicker.Stop()
+
+	// Start worker processors
+	for range config.concurrency {
+		go startWorkerProcessor(ctx, taskCh, respCh, rateLimiter, sdkClient, a.FrontendClient, metricsHandler, logger)
+	}
+
+	// Initialize the first p from initial executions or fetch from query
+	var p *page
+	if len(config.initialExecutions) > 0 {
+		// Use initial executions
+		p = &page{
+			executions:    config.initialExecutions,
+			nextPageToken: config.initialPageToken,
+			pageNumber:    hbd.CurrentPage,
+		}
+	} else {
+		// Fetch page of executions if needed
+		var err error
+		p, err = fetchPage(ctx, sdkClient, config, config.initialPageToken, hbd.CurrentPage)
+		if err != nil {
+			metrics.BatcherOperationFailures.With(metricsHandler).Record(1)
+			return HeartBeatDetails{}, fmt.Errorf("failed to fetch next page: %w", err)
+		}
+	}
+
+	for {
+		// Check if we need to fetch next page
+		if p.hasNext() && p.allSubmitted() {
+			nextPage, err := fetchPage(ctx, sdkClient, config, p.nextPageToken, p.pageNumber+1)
+			if err != nil {
+				metrics.BatcherOperationFailures.With(metricsHandler).Record(1)
+				return HeartBeatDetails{}, fmt.Errorf("failed to fetch next page: %w", err)
+			}
+			p.next = nextPage
+			nextPage.prev = p
+			p = nextPage
+
+			hbd.CurrentPage = p.pageNumber
+			hbd.PageToken = p.nextPageToken
+		}
+
+		select {
+		case taskCh <- p.nextTask():
+			p.submittedCount++
+
+		case result := <-respCh:
+			resultPage := result.page
+			if result.err == nil {
+				resultPage.successCount++
+			} else {
+				resultPage.errorCount++
+			}
+
+			// Update heartbeat details if this page and all previous pages are complete
+			if resultPage.done() {
+				allPrevComplete := true
+				for curr := resultPage.prev; curr != nil; curr = curr.prev {
+					if !curr.done() {
+						allPrevComplete = false
+						break
+					}
+				}
+				if allPrevComplete {
+					hbd.SuccessCount += resultPage.successCount
+					hbd.ErrorCount += resultPage.errorCount
+
+					activity.RecordHeartbeat(ctx, hbd)
+
+					resultPage.prev = nil
+				}
```

Comment:
I mentioned this in the previous review, `done()` is recursive already.

```suggestion
				hbd.SuccessCount += resultPage.successCount
				hbd.ErrorCount += resultPage.errorCount
```